src/tactic/norm_cast.lean

```diff
@@ -25,225 +25,314 @@ try_for 1000 (mk_instance e)
 end tactic
 
 namespace expr
+
 open tactic expr
 
 meta def flip_eq (ty : expr) : tactic (expr × (expr → expr)) :=
 do
-    (a, b) ← is_eq ty,
-    α ← infer_type a,
-    new_ty ← to_expr ``(%%b = %%a),
-    f ← to_expr ``(@eq.symm %%α %%a %%b),
-    return (new_ty, ⇑f)
+  (a, b) ← is_eq ty,
+  α ← infer_type a,
+  new_ty ← to_expr ``(%%b = %%a),
+  f ← to_expr ``(@eq.symm %%α %%a %%b),
+  return (new_ty, ⇑f)
 
 meta def flip_iff (ty : expr) : tactic (expr × (expr → expr)) :=
 do
-    (a, b) ← is_iff ty,
-    new_ty ← to_expr ``(%%b ↔ %%a),
-    f ← to_expr ``(@iff.symm %%a %%b),
-    return (new_ty, ⇑f)
+  (a, b) ← is_iff ty,
+  new_ty ← to_expr ``(%%b ↔ %%a),
+  f ← to_expr ``(@iff.symm %%a %%b),
+  return (new_ty, ⇑f)
+
+protected meta def to_pos_num : expr → option pos_num
+| `(@has_one.one %%α %%h) := some $ pos_num.one
+| `(@bit0 %%α %%h %%e) := do n ← to_pos_num e, return $ pos_num.bit0 n
+| `(@bit1 %%α %%h1 %%h2 %%e) := do n ← to_pos_num e, return $ pos_num.bit1 n
+| _ := none
+
+protected meta def to_num : expr → option num
+| `(@has_zero.zero %%α %%h) := some $ num.zero
+| e := do n ← e.to_pos_num, return $ num.pos n
 
 end expr
 
 namespace norm_cast
+
 open tactic expr
 
-private meta def new_name (n : name) : name := name.mk_string "reversed" n
+meta def pexpr_of_pos_num (α : expr) : pos_num → pexpr
+| pos_num.one := ``(has_one.one %%α)
+| (pos_num.bit0 n) := ``(bit0 %%(pexpr_of_pos_num n))
+| (pos_num.bit1 n) := ``(bit1 %%(pexpr_of_pos_num n))
 
-private meta def aux_after_set (tac : expr → tactic (expr × (expr → expr))) :
-    expr → tactic (expr × (expr → expr))
-| (pi n bi d b) := do
-    uniq_n ← mk_fresh_name,
-    let b' := b.instantiate_var (local_const uniq_n n bi d),
-    (b', f) ← aux_after_set b',
-    return $ (
-        pi n bi d $ b'.abstract_local uniq_n,
-        λ e, lam n bi d $ ( f $ e (local_const uniq_n n bi d) ).abstract_local uniq_n
-    )
-| ty := tac ty
+meta def pexpr_of_num (α : expr) : num → pexpr
+| num.zero := ``(has_zero.zero %%α)
+| (num.pos n) := pexpr_of_pos_num α n
 
-mk_simp_attribute push_cast "The `push_cast` simp attribute uses `move_cast` lemmas in the \"forward\" direction,
-to move casts toward the leaf nodes of the expression."
-
-/-- Called after the `move_cast` attribute is applied to a declaration. -/
-private meta def after_set (decl : name) (prio : ℕ) (pers : bool) : tactic unit :=
-do
-    (declaration.thm n l ty e) ← get_decl decl | failed,
-    let tac := λ ty, (flip_eq ty <|> flip_iff ty),
-    (ty', f) ← aux_after_set tac ty,
-    let e' := task.map f e,
-    let n' := new_name n,
-    add_decl (declaration.thm n' l ty' e'),
-    simp_attr.push_cast.set decl () tt
+meta def expr_of_num (α : expr) : num → tactic expr
```

Comment:
Move to `tactic/core.lean` (at the top, near similar decls) and rename `num.to_expr`.

Reply:
there's already an `expr.of_nat` tactic in there, maybe I should name this one `expr.of_num`?

Reply:
Sure, that's good too.